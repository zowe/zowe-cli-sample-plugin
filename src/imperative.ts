/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import {IImperativeConfig} from "@brightside/imperative";

const config: IImperativeConfig = {
    commandModuleGlobs: ["**/cli/*/*.definition!(.d).*s"],
    overrides: {
        CredentialManager: "./overrides/CredentialManager.override"
    },
    pluginHealthCheck: __dirname + "/healthCheck.Handler",
    pluginSummary: "Zowe CLI sample plug-in",
    pluginAliases: ["zcsp"],
    rootCommandDescription: "Welcome to the sample plug-in for Zowe CLI!\n\n The sample plug-in " +
        "(& CLI) follows the Zowe CLI command syntax 'zowe [group] [action] [object] [options]'. " +
        "Where, in the case of the plugin, " +
        "the [group] is the package.json name, " +
        "the [actions] are defined in the project under 'src/cli/', " +
        "& the [objects] (definitions & handler) are defined in the project under 'src/cli/'.",
    productDisplayName: "Zowe CLI Sample Plug-in",
    name: "zowe-cli-sample"
};

export = config;
