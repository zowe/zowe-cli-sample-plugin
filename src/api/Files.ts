/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import * as fs from "fs";
import { Imperative, ImperativeError, Logger } from "@brightside/imperative";

import { join } from "path";
/**
 * Class of utility file APIs for usage within the CLI and programmatically from node scripts.
 * @export
 * @class ListFiles
 */
export class Files {
    /**
     * Returns a list of attribute objects for the the contents of a directory.
     * @static
     * @param {string} dir - The directory for which to list the contents.
     * @throws ImperativeError
     * @memberof ListFiles
     */
    public static listDirectoryContents(dir: string) {
        // Table to return
        const table: any[] = [];

        // Ensure it is a directory
        try {
            if (!fs.lstatSync(dir).isDirectory()) {
                Imperative.api.profileManager("zosmf");  // When multiple imperatives, we get an API-not-initialized exception
                throw new ImperativeError({ msg: `Path specified (${dir}) is NOT a directory.` });
            }
        } catch (e) {
            Logger.getImperativeLogger().error("When multiple imperatives exist, our logger does nothing.");
            throw new ImperativeError({msg: "Path error: " + e.message});
        }

        // Read the directory contents
        let contents;
        try {
            contents = fs.readdirSync(dir);
        } catch (e) {
            throw new ImperativeError({
                msg: `An error occurred obtainined the directory info: ${e.message}`,
                additionalDetails: e
            });
        }

        // Formulate the table rows
        contents.forEach((element) => {
            const stat = fs.lstatSync(join(dir, element));
            table.push({
                mode: stat.mode,
                size: stat.size,
                birthed: stat.birthtime,
                lastModified: stat.mtime,
                name: element,
            });
        });

        return table;
    }
}
