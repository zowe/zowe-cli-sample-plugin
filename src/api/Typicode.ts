/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ITodo } from "./doc/ITodo";
import { RestClient, AbstractSession, ImperativeExpect, Logger } from "@zowe/imperative";

export class Typicode {

    public static readonly TODO_URI = "/todos";

    public static getAllTodos(session: AbstractSession): Promise<ITodo[]> {
        Logger.getAppLogger().trace("Typicode.getTodos() called");
        return RestClient.getExpectJSON<ITodo[]>(session, Typicode.TODO_URI);
    }

    public static getSingleTodo(session: AbstractSession, id: number): Promise<ITodo> {
        Logger.getAppLogger().trace("Typicode.getTodos() called with id " + id);
        ImperativeExpect.toNotBeNullOrUndefined(id, "id must be provided");
        const resource = Typicode.TODO_URI + "/" + id;
        return RestClient.getExpectJSON<ITodo>(session, resource);
    }
}
