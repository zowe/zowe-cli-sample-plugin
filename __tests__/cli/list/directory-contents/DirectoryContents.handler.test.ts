/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import {CheckStatus, ZosmfSession} from "@brightside/core";
import {Files} from "../../../../src/api/Files";
import {IHandlerParameters} from "@brightside/imperative";
import * as DirectoryDefinition from "../../../../src/cli/list/directory-contents/DirectoryContents.definition";
import * as DirectoryHandler from "../../../../src/cli/list/directory-contents/DirectoryContents.handler";

jest.mock("../../../../src/api/Files");

process.env.FORCE_COLOR = "0";

const DEFAULT_PARAMTERS: IHandlerParameters = {
    arguments: {
        $0: "bright",
        _: ["brightside-sample-plugin", "list", "directory-contents"],
    },
    profiles: {
        get: (type: string) => {
            return {};
        }
    } as any,
    response: {
        data: {
            setMessage: jest.fn((setMsgArgs) => {
                expect("" + setMsgArgs).toMatchSnapshot();
            }),
            setObj: jest.fn((setObjArgs) => {
                expect(setObjArgs).toMatchSnapshot();
            })
        },
        console: {
            log: jest.fn((logs) => {
                expect("" + logs).toMatchSnapshot();
            }),
            error: jest.fn((errors) => {
                expect("" + errors).toMatchSnapshot();
            }),
            errorHeader: jest.fn(() => undefined)
        },
        progress: {
            startBar: jest.fn((parms) => undefined),
            endBar: jest.fn(() => undefined)
        }
    } as any,
    definition: DirectoryDefinition.DirectoryContentsDefinition,
    fullDefinition: DirectoryDefinition.DirectoryContentsDefinition,
};

describe("directory-contents Handler", () => {
    afterEach(() => {
        jest.resetAllMocks();
    });
    it("should read a mocked directory", async () => {
        const fakeDir: string = "fake/directory";
        // Mock the files api call
        Files.listDirectoryContents = jest.fn((dir) => {
            return [{
                fake: "file",
                moreFake: "directory"
            }];
        });
        CheckStatus.getZosmfInfo = jest.fn(() => {
            return {zosmf_hostname: "dummy"};
        });
        ZosmfSession.createBasicZosmfSession = jest.fn();
        const handler = new DirectoryHandler.default();
        const params = Object.assign({}, ...[DEFAULT_PARAMTERS]);
        params.arguments.directory = fakeDir;

        // The handler should succeed
        await handler.process(params);

        expect(Files.listDirectoryContents).toHaveBeenCalledWith(fakeDir);
    });
});
