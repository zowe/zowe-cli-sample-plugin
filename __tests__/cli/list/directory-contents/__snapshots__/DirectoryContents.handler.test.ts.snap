// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`directory-contents Handler should read a mocked directory 1`] = `
"We just got a valid z/OSMF status response from system = dummy
"
`;

exports[`directory-contents Handler should read a mocked directory 2`] = `"Listing contents of \\"fake/directory\\""`;

exports[`directory-contents Handler should read a mocked directory 3`] = `
Array [
  Object {
    "fake": "file",
    "moreFake": "directory",
  },
]
`;

exports[`directory-contents Handler should read a mocked directory 4`] = `
"fake moreFake 
file directory"
`;
